src/chainstate/stacks/db/blocks.rs

```diff
@@ -0,0 +1,2427 @@
+                panic!("BUG: out-of-sequence microblock stream");
+            }
+            let cur_seq = (signed_microblocks[i-1].header.sequence as u32) + 1;
+            if cur_seq < (signed_microblocks[i].header.sequence as u32) {
+                // discontiguous
+                debug!("Discontiguous stream -- {} < {}", cur_seq, signed_microblocks[i].header.sequence);
+                return None;
+            }
+        }
+
+        // sanity check -- all parent block hashes are unique.  If there are duplicates, then the
+        // miner equivocated.
+        let mut parent_hashes : HashMap<BlockHeaderHash, StacksMicroblockHeader> = HashMap::new();
+        for i in 0..signed_microblocks.len() {
+            let signed_microblock = &signed_microblocks[i];
+            if parent_hashes.contains_key(&signed_microblock.header.prev_block) {
+                debug!("Deliberate microblock fork: duplicate parent {}", signed_microblock.header.prev_block.to_hex());
+                let conflicting_microblock_header = parent_hashes.get(&signed_microblock.header.prev_block).unwrap();
+
+                return Some((i - 1, Some(TransactionPayload::PoisonMicroblock(signed_microblock.header.clone(), conflicting_microblock_header.clone()))));
+            }
+            parent_hashes.insert(signed_microblock.header.prev_block.clone(), signed_microblock.header.clone());
+        }
+
+        // hashes are contiguous enough -- for each seqnum, there is a block with seqnum+1 with the
+        // block at seqnum as its parent.  There may be more than one.
+        for i in 1..signed_microblocks.len() {
+            if signed_microblocks[i - 1].header.sequence == signed_microblocks[i].header.sequence && signed_microblocks[i - 1].block_hash() != signed_microblocks[i].block_hash() {
+                // deliberate microblock fork
+                debug!("Deliberate microblock fork at sequence {}", signed_microblocks[i-1].header.sequence);
+                return Some((i - 1, Some(TransactionPayload::PoisonMicroblock(signed_microblocks[i - 1].header.clone(), signed_microblocks[i].header.clone()))));
+            }
+
+            if signed_microblocks[i - 1].block_hash() != signed_microblocks[i].header.prev_block {
+                // discontiguous
+                debug!("Discontinuous stream -- blocks not linked by hash");
+                return None;
+            }
+        }
+        
+        if anchored_block_header.parent_microblock == BlockHeaderHash([0u8; 32]) && anchored_block_header.parent_microblock_sequence == 0 {
+            // expected empty
+            debug!("Empty microblock stream between {} and {}", parent_anchored_block_header.block_hash().to_hex(), anchored_block_header.block_hash().to_hex());
+            return Some((0, None));
+        }
+
+        let num_microblocks = microblocks.len();
+
+        let mut end = 0;
+        let mut connects = false;
+        for i in 0..num_microblocks {
```

Comment:
should this be iterating over the signed_microblocks ?

Reply:
yes; will fix

Reply:
Fixed in b540a484ee9e610559b3a318f2946db2ecb7d6ce